opencensus/trace/trace_state.py

```diff
@@ -0,0 +1,98 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import OrderedDict
+import re
+
+_DELIMITER_FORMAT = '[ \t]*,[ \t]*'
+_KEY_FORMAT = r'[a-z][_0-9a-z\-\*\/]{0,255}'
+_VALUE_FORMAT = r'[\x20-\x2b\x2d-\x3c\x3e-\x7e]{1,256}'
+_MEMBER_FORMAT = '(%s)(=)(%s)' % (_KEY_FORMAT, _VALUE_FORMAT)
+
+_DELIMITER_FORMAT_RE = re.compile(_DELIMITER_FORMAT)
+_KEY_FORMAT_RE = re.compile(_KEY_FORMAT)
+_KEY_VALIDATION_RE = re.compile('^' + _KEY_FORMAT + '$')
+_VALUE_FORMAT_RE = re.compile(_VALUE_FORMAT)
+_VALUE_VALIDATION_RE = re.compile('^' + _VALUE_FORMAT + '$')
+_MEMBER_FORMAT_RE = re.compile(_MEMBER_FORMAT)
+
+
+class TraceState(OrderedDict):
+    HEADER_NAME = 'tracestate'
+
+    def __init__(self, *args, **kwds):
+        if len(args) == 1 and isinstance(args[0], str) and not kwds:
+            super(TraceState, self).__init__()
+            self.from_string(args[0])
+            return
+        super(TraceState, self).__init__(*args, **kwds)
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __setitem__(self, key, value):
+        if not isinstance(key, str):
+            raise ValueError('key must be an instance of str')
+        if not re.match(_KEY_VALIDATION_RE, key):
+            raise ValueError('illegal key provided')
+        if not isinstance(value, str):
+            raise ValueError('value must be an instance of str')
+        if not re.match(_VALUE_VALIDATION_RE, value):
+            raise ValueError('illegal value provided')
+        super(TraceState, self).__setitem__(key, value)
+
+    def __str__(self):
+        return 'TraceState(%r)' % (self.to_string())
+
+    def from_string(self, string):
+        for member in re.split(_DELIMITER_FORMAT_RE, string):
+            match = _MEMBER_FORMAT_RE.match(member)
+            if not match:
+                raise ValueError('illegal key-value format %r' % (member))
+            key, eq, value = match.groups()
+            self[key] = value
+        return self
+
+    def to_string(self):
+        return ','.join(map(lambda key: key + '=' + self[key], self))
+
+    def append(self, key, value):
+        if self.get(key):
+            del self[key]
+        self[key] = value
+
+    # make this an optional choice instead of enforcement during put/update
+    # if the tracestate value size is bigger than 512 characters, the tracer
+    # CAN decide to forward the tracestate
+    def is_valid(self):
+        if len(self) is 0:
+            return False
+        # combined header length MUST be less than or equal to 512 bytes
+        if len(self.to_string()) > 512:
+            return False
+        # there can be a maximum of 128 list-members in a list
+        if len(self) > 128:  # pragma: NO COVER
+            # it is impossible to pass the 512 length test and reach here ...
```

Comment:
I'll change 128 to 32 in the next PR (according to Bogdan's change: https://github.com/w3c/distributed-tracing/commit/17026e7836c6a08d537a2ebe65b22c290ec1e32f)

Reply:
@bogdandrutu would you grant me permission to trigger CI? Looks like the transient failure happened again. I can also look into the CI improvement if there is a need.

Reply:
@reyang I triggered it for you

Reply:
yes, I think there is a need =)